### PR TITLE
lavfi/dnn: Unification of Async and Sync Modes

### PR DESCRIPTION
**Pull request by shubhanshu02**
### Patch Set Description

This patchset is a part of optional deliverables in the GSoC project [Async Support for TensorFlow Backend in FFmpeg](https://summerofcode.withgoogle.com/projects/#5224576843251712).

- Builds on the patchset from PR #407. 

**Objective**: Unification of async and sync modes in the DNN backends from the filters' perspective.
**Parts under this deliverable**:
    - Unifies the async and sync modes in the DNN backend from the filters' perspective.
    - Adds `TaskItem` based inference in Native backend.
    - Renames `InferenceItem` to `LastLevelTaskItem`.

### Methodology

In this patchset, we plan to deliver a common inference function for async and sync modes to the filters (as compared to the two functions `ff_dnn_execute_model_async` and `ff_dnn_execute_model`). This will simplify the code on the filter side and the filters don't have to care about the mode of execution. With this patchset, the image frames are sent first by the frames and then collected using the `ff_dnn_get_result` function which sends back the input and output frames in a sequential manner.

The execution mode can now be set by using the `async` flag in the `backend_configs` as follows:
- async mode: `backend_configs=async=1`
- sync mode: `backend_configs=async=0`

Another part of this patchset contains the renaming of the `InferenceItem` in the three backends to `LastLevelTaskItem` and the relevant changes to avoid further confusion in the meaning of these structs.

### Patches

2c2a825b067c990f6af78e80316b484b5ae5b1aa Task-based Inference in Native Backend
5db57894287f23894d2db37d2e8745dc8da11dff Unify Execution Modes in DNN Filters
c0b4154e04d2da26116cd00d320183cb2cc6d9e9 Remove synchronous functions from DNN filters
42ce1db6f37eadb130b3426847ee28177e52fdcd Remove Async Flag from DNN Filter Side
5f0a0ed7ed3d11af31f831f237cff916341dd4b8 Rename InferenceItem to LastLevelTaskItem
85fed1e447bd121089acc623999e1a712cc938bd Include dnn_processing in docs of sr and derain filter

### CODE REVIEW

File: libavfilter/dnn/dnn_interface.c [hunk 1]
```diff
@@ -33,6 +33,10 @@ DNNModule *ff_get_dnn_module(DNNBackendType backend_type)
 {
     DNNModule *dnn_module;
 
+    // temporary change till unification completes in atleast one backend
+    av_log(NULL, AV_LOG_ERROR, "Execution currently not supported \n");
```
Comment (guoyejun) [inline libavfilter/dnn/dnn_interface.c]: it's not good to disable a feature even it is temporary, we need to refine the patches

Comment (guoyejun) [reply libavfilter/dnn/dnn_interface.c]: it might be a good method for close/internal code, but it is not good for open source.

Comment (shubhanshu02) [reply libavfilter/dnn/dnn_interface.c]: Okay, got it.
How about reordering the "Task-based Inference in Native Backend" to first and then combining the unification patches in the three backend to one?

Comment (guoyejun) [reply libavfilter/dnn/dnn_interface.c]: looks good, hope the merged patch is not big, :)

Comment (shubhanshu02) [reply libavfilter/dnn/dnn_interface.c]: I'll try not to make the patch too large. Once refining is complete, I'll tell you. Thank you.

Comment (shubhanshu02) [reply libavfilter/dnn/dnn_interface.c]: I tried merging the patches. The first two patches are a bit large, but the next two are small.

File: libavfilter/dnn/dnn_backend_native.c [hunk 2]
```diff
@@ -45,9 +45,29 @@ static const AVClass dnn_native_class = {
     .category   = AV_CLASS_CATEGORY_FILTER,
 };
 
-static DNNReturnType execute_model_native(const DNNModel *model, const char *input_name, AVFrame *in_frame,
-                                          const char **output_names, uint32_t nb_output, AVFrame *out_frame,
-                                          int do_ioproc);
+static DNNReturnType execute_model_native(Queue *inference_queue);
+
+static DNNReturnType extract_inference_from_task(TaskItem *task, Queue *inference_queue)
```
Comment (guoyejun) [inline libavfilter/dnn/dnn_backend_native.c]: fyi, we may change the inference in a new patch after this patch set is finished, it is really confusing.

the struct name can be LastLevelTaskItem, and the function name may extract_lltask_from_task, we use lltask for variable.

The last level task means the task can be passed to the inference engine, so, 'task' for the interface between filter and ffmpeg-dnn, while lltask is the interface between dnn and inference engine.

File: libavfilter/dnn/dnn_backend_native.c [hunk 3]
```diff
@@ -34,6 +34,7 @@
 #define FLAGS AV_OPT_FLAG_FILTERING_PARAM
 static const AVOption dnn_native_options[] = {
     { "conv2d_threads", "threads num for conv2d layer", OFFSET(options.conv2d_threads), AV_OPT_TYPE_INT,  { .i64 = 0 }, INT_MIN, INT_MAX, FLAGS },
+    { "async",          "use DNN async inference",      OFFSET(options.async),          AV_OPT_TYPE_BOOL, { .i64 = 1 },       0,       1, FLAGS },
```
Comment (guoyejun) [inline libavfilter/dnn/dnn_backend_native.c]: looks that the code in dnn_backend_native.c/h can be a separate patch, as the preparation for the async/sync unification of filter.

File: libavfilter/dnn/dnn_backend_openvino.c [hunk 4]
```diff
@@ -375,6 +376,11 @@ static DNNReturnType init_model_ov(OVModel *ov_model, const char *input_name, co
         ctx->options.nireq = av_cpu_count() / 2 + 1;
     }
 
+    if (ctx->options.async != 0 && ctx->options.async != 1) {
```
Comment (guoyejun) [inline libavfilter/dnn/dnn_backend_openvino.c]: do we need this if? The allowed values are 0 and 1.

    { "async",           "use DNN async inference",       OFFSET(options.async),           AV_OPT_TYPE_BOOL,   { .i64 = 1 },     0,       1, FLAGS },

File: libavfilter/dnn/dnn_backend_openvino.c [hunk 5]
```diff
@@ -724,6 +730,8 @@ DNNModel *ff_dnn_load_model_ov(const char *model_filename, DNNFunctionType func_
 
     //parse options
     av_opt_set_defaults(ctx);
+    // set default async to 1
+    ctx->options.async = 1;
```
Comment (guoyejun) [inline libavfilter/dnn/dnn_backend_openvino.c]: .i64=1 in dnn_backend_common.h  has set the default value as 1.

File: libavfilter/dnn/dnn_backend_tf.c [hunk 6]
```diff
@@ -83,6 +84,7 @@ typedef struct TFRequestItem {
 #define FLAGS AV_OPT_FLAG_FILTERING_PARAM
 static const AVOption dnn_tensorflow_options[] = {
     { "sess_config", "config for SessionOptions", OFFSET(options.sess_config), AV_OPT_TYPE_STRING, { .str = NULL }, 0, 0, FLAGS },
+    { "async",  "use DNN async inference",OFFSET(options.async), AV_OPT_TYPE_BOOL, { .i64 = 1 }, 0, 1, FLAGS},
```
Comment (guoyejun) [inline libavfilter/dnn/dnn_backend_tf.c]: need this line?

File: libavfilter/dnn/dnn_backend_tf.c [hunk 7]
```diff
@@ -868,6 +870,8 @@ DNNModel *ff_dnn_load_model_tf(const char *model_filename, DNNFunctionType func_
 
     //parse options
     av_opt_set_defaults(ctx);
+    // set default async to 1
+    ctx->options.async = 1;
```
Comment (guoyejun) [inline libavfilter/dnn/dnn_backend_tf.c]: need these 2 lines?

File: libavfilter/dnn/dnn_backend_tf.c [hunk 8]
```diff
@@ -883,6 +887,10 @@ DNNModel *ff_dnn_load_model_tf(const char *model_filename, DNNFunctionType func_
         ctx->options.nireq = av_cpu_count() / 2 + 1;
     }
 
+    if (ctx->options.async != 0 && ctx->options.async != 1) {
```
Comment (guoyejun) [inline libavfilter/dnn/dnn_backend_tf.c]: these

File: libavfilter/vf_dnn_detect.c [hunk 9]
```diff
@@ -539,5 +539,5 @@ const AVFilter ff_vf_dnn_detect = {
     .inputs        = dnn_detect_inputs,
     .outputs       = dnn_detect_outputs,
     .priv_class    = &dnn_detect_class,
-    .activate      = dnn_detect_activate,
+    .activate      = dnn_detect_activate_async,
```
Comment (guoyejun) [inline libavfilter/vf_dnn_detect.c]: still use dnn_detect_activate here.  It is not good to see 'async' or 'sync' here, since the filter will not care about this concept.  For less change, we may change the function name from dnn_detect_activate_async into dnn_detect_activate_impl, and call dnn_detect_activate_impl in function dnn_detect_activate.  and then delete function  dnn_detect_activate_sync.

```
static int dnn_detect_activate(AVFilterContext *filter_ctx)
{
    DnnDetectContext *ctx = filter_ctx->priv;

    if (ctx->dnnctx.async)  // do not remove this variable, but to add deprecated, refer to 'options' in dnn_filter_common.h
        output warning here to be backward compatible.
   
    return dnn_detect_activate_impl(filter_ctx);
}
```
the same comment for dnn_processing.

Comment (guoyejun) [reply libavfilter/vf_dnn_detect.c]: ok, please ignore this comment, since I see it is fixed in the next patch.

File: libavfilter/dnn/dnn_backend_tf.c [hunk 10]
```diff
@@ -891,6 +891,13 @@ DNNModel *ff_dnn_load_model_tf(const char *model_filename, DNNFunctionType func_
         ctx->options.async = 1;
     }
 
+#if !HAVE_PTHREAD_CANCEL
+    if (ctx->async) {
```
Comment (guoyejun) [inline libavfilter/dnn/dnn_backend_tf.c]: ctx->options.async?

File: libavfilter/dnn_filter_common.h [hunk 11]
```diff
@@ -46,7 +45,6 @@ typedef struct DnnContext {
     { "output",             "output name of the model",   OFFSET(model_outputnames_string), AV_OPT_TYPE_STRING, { .str = NULL }, 0, 0, FLAGS },\
     { "backend_configs",    "backend configs",            OFFSET(backend_options),  AV_OPT_TYPE_STRING,    { .str = NULL }, 0, 0, FLAGS },\
     { "options", "backend configs (deprecated, use backend_configs)", OFFSET(backend_options),  AV_OPT_TYPE_STRING, { .str = NULL }, 0, 0, FLAGS | AV_OPT_FLAG_DEPRECATED},\
-    { "async",              "use DNN async inference",    OFFSET(async),            AV_OPT_TYPE_BOOL,      { .i64 = 1},     0, 1, FLAGS},
```
Comment (guoyejun) [inline libavfilter/dnn_filter_common.h:49]: do not remove it for backward compatible, just add comment such as: ignored, use backend_configs='async=1' ...

Comment (guoyejun) [reply libavfilter/dnn_filter_common.h:49]: I mean comment updated in the second column, "use DNN async inference"

File: libavfilter/dnn/dnn_backend_openvino.c [hunk 12]
```diff
@@ -846,23 +806,46 @@ DNNReturnType ff_dnn_execute_model_async_ov(const DNNModel *model, DNNExecBasePa
...
+                return ret;
+            }
+        }
+
+        return DNN_SUCCESS;
+    }
+    else {
+        if (model->func_type == DFT_ANALYTICS_CLASSIFY) {
+            // Once we add async support for tensorflow backend and native backend,
+            // we'll combine the two sync/async functions in dnn_interface.h to
```
Comment (guoyejun) [inline libavfilter/dnn/dnn_backend_openvino.c]: not catch this point.

we do support asyn for tf backend now.
and sync/async is combined in next patch without any change here.

I know that classification is not supported/tested for sync in the previous, is this the root cause? So we return error here to first keep the same feature as previous.

File: libavfilter/dnn/dnn_backend_openvino.c [hunk 13]
```diff
@@ -846,23 +806,46 @@ DNNReturnType ff_dnn_execute_model_async_ov(const DNNModel *model, DNNExecBasePa
...
+            // simplify the code in filter, and async will be an option within backends.
+            // so, do not support now, and classify filter will not call this function.
+            return DNN_ERROR;
+        }
+
+        if (ctx->options.batch_size > 1) {
+            avpriv_report_missing_feature(ctx, "batch mode for sync execution");
```
Comment (guoyejun) [inline libavfilter/dnn/dnn_backend_openvino.c:827]: looks that batch mode for sync execution can be supported without much effort, right?  I do not mean you'd add such support or not, just a discussion.

File: libavfilter/dnn/dnn_backend_native.c (hunk 2 above)
Comment (shubhanshu02) [reply libavfilter/dnn/dnn_backend_native.c]: Sure, got the point. Shall I include this in this patchset as the last patch or include it in the next patch set?

File: libavfilter/dnn/dnn_backend_openvino.c (hunk 4 above)
Comment (shubhanshu02) [reply libavfilter/dnn/dnn_backend_openvino.c]: Oh right. Thanks for the catch.

File: libavfilter/dnn/dnn_backend_openvino.c (hunk 5 above)
Comment (shubhanshu02) [reply libavfilter/dnn/dnn_backend_openvino.c]: Oh so this sets the default value. Okay I'll remove this line

File: libavfilter/dnn/dnn_backend_openvino.c (hunk 13 above)
Comment (shubhanshu02) [reply libavfilter/dnn/dnn_backend_openvino.c:827]: Yes. It seems like we can use the same code we do for async batch mode to support this. Just that we'd need to be sure that the flush frame function is being called for edge cases.

File: libavfilter/dnn/dnn_backend_native.c (hunk 3 above)
Comment (shubhanshu02) [reply libavfilter/dnn/dnn_backend_native.c]: like which part? The conversion to `TaskItem` based inference is in a separate patch.

File: libavfilter/vf_dnn_detect.c (hunk 9 above)
Comment (shubhanshu02) [reply libavfilter/vf_dnn_detect.c]: Sure.

File: libavfilter/dnn/dnn_backend_tf.c (hunk 10 above)
Comment (shubhanshu02) [reply libavfilter/dnn/dnn_backend_tf.c]: Yes, corrected.

File: libavfilter/dnn_filter_common.h (hunk 11 above)
Comment (shubhanshu02) [reply libavfilter/dnn_filter_common.h:49]: Like this "use DNN async inference (ignored, use backend_configs='async=1')"?

File: libavfilter/dnn/dnn_backend_openvino.c (hunk 12 above)
Comment (shubhanshu02) [reply libavfilter/dnn/dnn_backend_openvino.c]: Yes. I checked with the following two filters in this way and found them to be working with sync mode as well.

```sh
ffmpeg -i cici.jpg -vf dnn_classify=dnn_backend=openvino:model=emotions-recognition-retail-0003.xml:input=data:output=prob_emotion:confidence=0.3:labels=emotions-recognition-retail-0003.label:target=face:backend_configs=async=0,showinfo -f null -
```

```sh
ffmpeg -i cici.jpg -vf dnn_detect=dnn_backend=openvino:model=face-detection-adas-0001.xml:input=data:output=detection_out:confidence=0.6:labels=face-detection-adas-0001.label:backend_configs=async=0,showinfo -f null -
```

But I did not remove these lines as I wasn't sure if it will work in all cases. Like if there are other cases as well.

File: libavfilter/dnn/dnn_backend_openvino.c (hunk 13 above)
Comment (guoyejun) [reply libavfilter/dnn/dnn_backend_openvino.c:827]: i see

File: libavfilter/dnn_filter_common.h (hunk 11 above)
Comment (guoyejun) [reply libavfilter/dnn_filter_common.h:49]: yes, otherwise, people will see program reports error and ends due to the unknown option, if he used async in the command line.

File: libavfilter/dnn/dnn_backend_native.c (hunk 3 above)
Comment (guoyejun) [reply libavfilter/dnn/dnn_backend_native.c]: i meant the code in dnn_backend_native.c/h in the second patch. anyway, it doesn't matter to separate it.

File: libavfilter/dnn/dnn_backend_native.c (hunk 2 above)
Comment (guoyejun) [reply libavfilter/dnn/dnn_backend_native.c]: add as the last patch would be better

File: libavfilter/dnn/dnn_backend_openvino.c (hunk 12 above)
Comment (guoyejun) [reply libavfilter/dnn/dnn_backend_openvino.c]: the first command line is not correct, dnn_classify will do nothing since no detect result, and the second command line just contain dnn_detect, not dnn_classify.
To use classify, we need to use the two filters together:  -vf dnn_detect=...dnn_classify=...

my concern was about the comment, we may update the comment to be more clear, keeping return error.

File: libavfilter/dnn/dnn_backend_native.c (hunk 2 above)
Comment (shubhanshu02) [reply libavfilter/dnn/dnn_backend_native.c]: Okay, I'll add the patch today.

File: libavfilter/dnn_filter_common.h (hunk 11 above)
Comment (shubhanshu02) [reply libavfilter/dnn_filter_common.h:49]: Done. I had a question - how will people who haven't seen the source code see this comment? Is there any help command for the filter?

File: libavfilter/dnn/dnn_backend_openvino.c (hunk 12 above)
Comment (shubhanshu02) [reply libavfilter/dnn/dnn_backend_openvino.c]: Oh yes, I just noticed it didn't return any coordinates of the regions classified.

How about this?
```
        // Classification filter has not been completely tested with the sync
        // mode and on the TensorFlow and Native backends. So, do not support now
```

File: libavfilter/dnn/dnn_backend_openvino.c (hunk 13 above)
Comment (shubhanshu02) [reply libavfilter/dnn/dnn_backend_openvino.c:827]: I tried it also and it works fine since all filters which can use `batch_mode` call the flush frame function.

File: libavfilter/dnn/dnn_backend_native.c (hunk 3 above)
Comment (shubhanshu02) [reply libavfilter/dnn/dnn_backend_native.c]: Okay

File: libavfilter/dnn_filter_common.h (hunk 11 above)
Comment (guoyejun) [reply libavfilter/dnn_filter_common.h:49]: yes, it is not complete backward compatible.

ffmpeg -h filter=<name of filter>

File: libavfilter/dnn/dnn_backend_openvino.c (hunk 12 above)
Comment (guoyejun) [reply libavfilter/dnn/dnn_backend_openvino.c]: just the first line is enough.

we can always support here no matter if tf/native backends are supported or not,

File: libavfilter/dnn/dnn_backend_openvino.c (hunk 13 above)
Comment (guoyejun) [reply libavfilter/dnn/dnn_backend_openvino.c:827]: we may enable it in another new patch some time later

File: libavfilter/dnn/dnn_backend_openvino.c (hunk 12 above)
Comment (shubhanshu02) [reply libavfilter/dnn/dnn_backend_openvino.c]: Okay. Updated.

File: libavfilter/dnn/dnn_backend_openvino.c (hunk 13 above)
Comment (shubhanshu02) [reply libavfilter/dnn/dnn_backend_openvino.c:827]: Sure

File: libavfilter/dnn_filter_common.h (hunk 11 above)
Comment (shubhanshu02) [reply libavfilter/dnn_filter_common.h:49]: I was thinking that we can make it backward compatible by passing this async parameter to the backend load function. 

Since this `async` is present in `DnnContext`, we can pass it to the load model functions and check if the user has set the value there. (in this case, we'd have to set a default value other than 0 or 1)

The only problem with this method is that if the user sets both the values `async=0` and `backend_configs=async=1`, we'll not know which one user was trying to use.

File: libavfilter/dnn/dnn_backend_native.c (hunk 2 above)
Comment (shubhanshu02) [reply libavfilter/dnn/dnn_backend_native.c]: What about the `inference_queue`? Shall I name it `lltask_queue`?

If I understood correctly, the request item should look like this, right?
```c
typedef struct OVRequestItem {
    ie_infer_request_t *infer_request;
    LastLevelTaskItem **lltasks;
    uint32_t lltask_count;
    ie_complete_call_back_t callback;
} OVRequestItem;
```

Comment (guoyejun) [reply libavfilter/dnn/dnn_backend_native.c]: yes. right.

File: libavfilter/dnn_filter_common.h (hunk 11 above)
Comment (guoyejun) [reply libavfilter/dnn_filter_common.h:49]: i guess not many people are using the filter option, and such  backward compatible is too complex ...

File: libavfilter/dnn/dnn_backend_native.c (hunk 2 above)
Comment (shubhanshu02) [reply libavfilter/dnn/dnn_backend_native.c]: Okay, thanks.

File: libavfilter/dnn_filter_common.h (hunk 11 above)
Comment (shubhanshu02) [reply libavfilter/dnn_filter_common.h:49]: Okay. I have changed the comment in column 2.

File: libavfilter/dnn/dnn_backend_native.c (hunk 2 above)
Comment (shubhanshu02) [reply libavfilter/dnn/dnn_backend_native.c]: Added

File: libavfilter/dnn/dnn_backend_native.c [hunk 14]
```diff
@@ -185,10 +186,16 @@ DNNModel *ff_dnn_load_model_native(const char *model_filename, DNNFunctionType f
 
     native_model->ctx.class = &dnn_native_class;
     model->options = options;
+    native_model->ctx.options.async = 0;
```
Comment (guoyejun) [inline libavfilter/dnn/dnn_backend_native.c]: the option has default value

File: libavfilter/dnn/dnn_backend_tf.c [hunk 15]
```diff
@@ -1156,7 +1156,7 @@ DNNReturnType ff_dnn_execute_model_tf(const DNNModel *model, DNNExecBaseParams *
     }
 
     if (extract_lltask_from_task(task, tf_model->inference_queue) != DNN_SUCCESS) {
-        av_log(ctx, AV_LOG_ERROR, "unable to extract inference from task.\n");
+        av_log(ctx, AV_LOG_ERROR, "unable to extract lltask from task.\n");
```
Comment (guoyejun) [inline libavfilter/dnn/dnn_backend_tf.c]: for log message, we can use the full name: last level task instead of lltask

**Comment by guoyejun** [comment]
the four patches relative to last level task can be merged together, I see it would be large (thanks for considering small patches), it is more nature to be together.

File: libavfilter/dnn/dnn_backend_native.c [hunk 16]
```diff
@@ -425,24 +437,57 @@ DNNReturnType ff_dnn_execute_model_native(const DNNModel *model, DNNExecBasePara
...
+
+    // for now, use sync node with flush operation
+    // Switch to async when it is supported
+    return execute_model_native(native_model->inference_queue);
+}
+
+DNNAsyncStatusType ff_dnn_get_result_native(const DNNModel *model, AVFrame **in, AVFrame **out)
+{
+    NativeModel *native_model = model->model;
+    return ff_dnn_get_async_result_common(native_model->task_queue, in, out);
```
Comment (guoyejun) [inline libavfilter/dnn/dnn_backend_native.c]: we need to change the function name to ff_dnn_get_result_common, up to you to decide which patch is best for the change.

File: libavfilter/dnn/dnn_backend_native.c (hunk 14 above)
Comment (shubhanshu02) [reply libavfilter/dnn/dnn_backend_native.c]: Corrected

File: libavfilter/dnn/dnn_backend_tf.c (hunk 15 above)
Comment (shubhanshu02) [reply libavfilter/dnn/dnn_backend_tf.c]: Sure

File: libavfilter/dnn/dnn_backend_native.c (hunk 16 above)
Comment (shubhanshu02) [reply libavfilter/dnn/dnn_backend_native.c]: I merged it in the unification patch.

**Comment by guoyejun** [comment]
looks good to me, please send to ffmpeg community. thanks.

@Semmer2  could you also have a test after the patches are sent, thanks.

**Comment by shubhanshu02** [comment]
Sure, I'll send it in a few minutes after rebasing to the latest master. Thank you for reviewing.

**Comment by guoyejun** [comment]
lgtm

**Comment by Semmer2** [comment]
Hi,
The new patches may crushed when using tf backend with dnn_detect. It may coused by null frame of dnn_detect_post_proc. Pleas check this.
The command I used is `./ffmpeg -i dog.jpg -vf dnn_detect=dnn_backend=tensorflow:input=image_tensor:output="num_detections&detection_scores&detection_classes&detection_boxes":model=ssd_mobilenet_v2_coco.pb:backend_configs='async=0',showinfo -y face_detect.jpeg`
Please re-check all the dnn filter and backends. :)

**Comment by shubhanshu02** [comment]
I have corrected the patch and tested it with the following commands.

#### Classification 
I tried with these values of async (0,0), (0,1), (1,0), (1,1) where (x,y) = (async value in detect, async value in classify)
```
ffmpeg -i cici.jpg -vf dnn_detect=dnn_backend=openvino:model=face-detection-adas-0001.xml:backend_configs='async=0':input=data:output=detection_out:confidence=0.6:labels=face-detection-adas-0001.label,dnn_classify=dnn_backend=openvino:model=emotions-recognition-retail-0003.xml:input=data:output=prob_emotion:confidence=0.3:labels=emotions-recognition-retail-0003.label:backend_configs='async=0',showinfo -f null -
```
#### Detection
Tried with `async=0` and `async=1`

```
ffmpeg -i cici.jpg -vf dnn_detect=dnn_backend=openvino:model=face-detection-adas-0001.xml:backend_configs='async=0':input=data:output=detection_out:confidence=0.6:labels=face-detection-adas-0001.label,showinfo -f null -
```
```
ffmpeg -i dog.jpg -vf  dnn_detect=dnn_backend=tensorflow:input=image_tensor:output="num_detections&detection_scores&detection_classes&detection_boxes":model=ssd_mobilenet_v2_coco.pb:backend_configs='async=1',showinfo -y face_detect.jpeg
```
#### Derain

Tried with `async=0` and `async=1`

```
ffmpeg -i rain.jpg -vf derain=model=can.pb:dnn_backend=1 derain.jpg -y 
```
```
ffmpeg -i rain.jpg -vf derain=model=can.model:dnn_backend=0 derain.jpg -y 
```

#### SR filter
Tried with `async=0` and `async=1`
```
ffmpeg -i 480p.jpg -vf sr=model=espcn.pb:dnn_backend=1 espcn.jpg -y 
```

```
ffmpeg -i 480p.jpg -vf sr=model=espcn.model:dnn_backend=0 espcn.jpg -y 
```

#### DNN Processing

```
ffmpeg -i rain.jpg -vf format=rgb24,dnn_processing=dnn_backend=openvino:model=can.xml:input=x:output=can/sub:backend_configs="input_resizable=1&async=0" -y rain1.jpg
```
```
ffmpeg -i rain.jpg -vf format=rgb24,dnn_processing=dnn_backend=openvino:model=can.xml:input=x:output=can/sub:backend_configs="input_resizable=1&async=1" -y rain1.jpg
```
```
ffmpeg -i rain.jpg -vf format=rgb24,dnn_processing=dnn_backend=tensorflow:model=can.pb:input=x:output=y:backend_configs='async=1' derain.jpg -y
```
```
ffmpeg -i rain.jpg -vf format=rgb24,dnn_processing=dnn_backend=tensorflow:model=can.pb:input=x:output=y:backend_configs='async=0' derain.jpg -y
```
```
ffmpeg -i rain.jpg -vf format=rgb24,dnn_processing=dnn_backend=native:model=can.model:input=x:output=y:backend_configs='async=0' derain.jpg -y
```
```
ffmpeg -i rain.jpg -vf format=rgb24,dnn_processing=dnn_backend=native:model=can.model:input=x:output=y:backend_configs='async=1' derain.jpg -y
```
In the native backend, async is not supported, so `async=1` ran in sync mode.
Can you have a look at the above commands and check if I missed any filter?

Thank you.

**Comment by shubhanshu02** [comment]
I have corrected the patch and tested it with the following commands.

#### Classification 
I tried with these values of async (0,0), (0,1), (1,0), (1,1) where (x,y) = (async value in detect, async value in classify)
```
ffmpeg -i cici.jpg -vf dnn_detect=dnn_backend=openvino:model=face-detection-adas-0001.xml:backend_configs='async=0':input=data:output=detection_out:confidence=0.6:labels=face-detection-adas-0001.label,dnn_classify=dnn_backend=openvino:model=emotions-recognition-retail-0003.xml:input=data:output=prob_emotion:confidence=0.3:labels=emotions-recognition-retail-0003.label:backend_configs='async=0',showinfo -f null -
```
#### Detection
Tried with `async=0` and `async=1`

```
ffmpeg -i cici.jpg -vf dnn_detect=dnn_backend=openvino:model=face-detection-adas-0001.xml:backend_configs='async=0':input=data:output=detection_out:confidence=0.6:labels=face-detection-adas-0001.label,showinfo -f null -
```
```
ffmpeg -i dog.jpg -vf  dnn_detect=dnn_backend=tensorflow:input=image_tensor:output="num_detections&detection_scores&detection_classes&detection_boxes":model=ssd_mobilenet_v2_coco.pb:backend_configs='async=1',showinfo -y face_detect.jpeg
```
#### Derain

Tried with `async=0` and `async=1`

```
ffmpeg -i rain.jpg -vf derain=model=can.pb:dnn_backend=1 derain.jpg -y 
```
```
ffmpeg -i rain.jpg -vf derain=model=can.model:dnn_backend=0 derain.jpg -y 
```

#### SR filter
Tried with `async=0` and `async=1`
```
ffmpeg -i 480p.jpg -vf sr=model=espcn.pb:dnn_backend=1 espcn.jpg -y 
```

```
ffmpeg -i 480p.jpg -vf sr=model=espcn.model:dnn_backend=0 espcn.jpg -y 
```

#### DNN Processing
Tried with `async=0` and `async=1`
```
ffmpeg -i rain.jpg -vf format=rgb24,dnn_processing=dnn_backend=openvino:model=can.xml:input=x:output=can/sub:backend_configs="input_resizable=1&async=0" -y rain1.jpg
```

```
ffmpeg -i rain.jpg -vf format=rgb24,dnn_processing=dnn_backend=tensorflow:model=can.pb:input=x:output=y:backend_configs='async=1' derain.jpg -y
```
```
ffmpeg -i rain.jpg -vf format=rgb24,dnn_processing=dnn_backend=native:model=can.model:input=x:output=y:backend_configs='async=0' derain.jpg -y
```

In the native backend, async is not supported, so `async=1` ran in sync mode.

@guoyejun @Semmer2 Sir, can you have a look at the above commands and check if I missed any filter? Thank you.

**Comment by Semmer2** [comment]
Hi Shubhanshu,
LGTM. And the code passed my local test. :)

**Comment by shubhanshu02** [comment]
Thanks. I'll send the updated patches to the mailing list.

**Comment by guoyejun** [comment]
we may need to merge the last patch into the patch which causes the issue

**Comment by shubhanshu02** [comment]
> we may need to merge the last patch into the patch which causes the issue

I think the patch `libavfilter: Unify Execution Modes in DNN Filters` is the one that caused this issue. When I build on it, it showed the same bug. The one just before it is related to the native backend, so won't be contributing to this problem.

Shall I merge these? (these =  `libavfilter: Unify Execution Modes in DNN Filters` + `libavfilter: Send only input frame for DNN Detect and Classify`)

**Comment by guoyejun** [comment]
> > we may need to merge the last patch into the patch which causes the issue
> 
> I think the patch `libavfilter: Unify Execution Modes in DNN Filters` is the one that caused this issue. When I build on it, it showed the same bug. The one just before it is related to the native backend, so won't be contributing to this problem.
> 
> Shall I merge these? (these = `libavfilter: Unify Execution Modes in DNN Filters` + `libavfilter: Send only input frame for DNN Detect and Classify`)

yes, we'd keep every single patch not breaking things.

**Comment by guoyejun** [comment]
i see you just force-pushed the patches, is it a rebase or something new?

**Comment by shubhanshu02** [comment]
> i see you just force-pushed the patches, is it a rebase or something new?

Actually, I saw some CI tests failing, so thought to rebase against the latest master. Nothing new added.

**Comment by shubhanshu02** [comment]
Thank you for reviewing @guoyejun @Semmer2 Sir.
